tests/test_base_virtual_machine_factory.cpp

```diff
@@ -121,3 +135,167 @@ TEST_F(BaseFactory, creates_cloud_init_iso_image)
+    const auto nets_copy = nets;
+
+    factory.prepare_networking(nets);
+    EXPECT_EQ(nets, nets_copy);
+}
+
+TEST_F(BaseFactory, prepareNetworkingGutsWithNoExtraNetsHasNoObviousEffect)
+{
+    StrictMock<MockBaseFactory> factory;
+
+    std::vector<mp::NetworkInterface> empty;
+    factory.base_prepare_networking_guts(empty, "asdf");
+    EXPECT_THAT(empty, IsEmpty());
+}
+
+TEST_F(BaseFactory, prepareNetworkingGutsLeavesUnrecognizedNetworksAlone)
+{
+    StrictMock<MockBaseFactory> factory;
+    EXPECT_CALL(factory, networks).WillOnce(Return(std::vector<mp::NetworkInterfaceInfo>{{"eth0", "ethernet", "asd"}}));
+
+    std::vector<mp::NetworkInterface> extra_nets{{"wlan0", "fa:se:ma:c0:12:23", false},
+                                                 {"eth1", "fa:se:ma:c9:45:56", true}};
+    const auto extra_copy = extra_nets;
+
+    factory.base_prepare_networking_guts(extra_nets, "bridge");
+    EXPECT_EQ(extra_nets, extra_copy);
+}
+
+TEST_F(BaseFactory, prepareNetworkingGutsLeavesBridgeTypeNetworksAlone)
+{
+    StrictMock<MockBaseFactory> factory;
+    constexpr auto bridge_type = "arbitrary";
+    EXPECT_CALL(factory, networks)
+        .WillOnce(
+            Return(std::vector<mp::NetworkInterfaceInfo>{{"br0", bridge_type, "foo"}, {"xyz", bridge_type, "bar"}}));
+
+    std::vector<mp::NetworkInterface> extra_nets{{"br0", "false mac", false}, {"xyz", "fake mac", true}};
+    const auto extra_copy = extra_nets;
+
+    factory.base_prepare_networking_guts(extra_nets, bridge_type);
+    EXPECT_EQ(extra_nets, extra_copy);
+}
+
+struct TestBridgePreparation
+    : public BaseFactory,
+      WithParamInterface<std::tuple<std::vector<mp::NetworkInterface>, std::vector<mp::NetworkInterface>,
+                                    std::vector<mp::NetworkInterface>>>
+{
+    inline static constexpr auto br_type = "tunnel";
+    inline static const std::vector<mp::NetworkInterfaceInfo> host_nets{
+        {"eth0", "ethernet", "e0"}, {"eth1", "ethernet", "e1"},       {"wlan0", "wifi", "w0"},
+        {"wlan1", "wifi", "w1"},    {"br0", br_type, "b0", {"eth0"}}, {"br1", br_type, "b1", {"wlan1"}},
+        {"br2", br_type, "empty"}};
+    std::vector<mp::NetworkInterface> mix_extra_nets() const // mixes all requested
+    {
+        std::vector<mp::NetworkInterface> ret;
+        const auto& [requested_bridged, requested_unbridged, requested_existing_bridges] = GetParam();
+
+        auto maxim = std::max(requested_bridged.size(), requested_unbridged.size());
+        maxim = std::max(maxim, requested_existing_bridges.size());
+        for (size_t i = 0; i < maxim; ++i) // alternate to avoid testing only partitioned extra_nets
+            for (const auto& requests : {requested_bridged, requested_unbridged, requested_existing_bridges})
+                if (i < requests.size())
+                    ret.push_back(requests[i]);
+
+        return ret;
+    }
+};
+
+TEST_P(TestBridgePreparation, prepareNetworkingGutsCreatesBridgesAndReplacesIds)
```

Comment:
This test a bit unwieldy, could split in the three/four cases it handles?

Reply:
Yeah it is indeed, and it was really hard to get right! It was approx. a full day's worth of grind, going different tests, then single, then parameter, then back and forth again, getting the matchers to work...

The problem is that we need to match against a container, with multiple elements, corresponding to different scenarios. It would be much simpler to test single elements, but the code could break on multiple. And if I test multiple, even with all elements corresponding to a single scenario (say all existing bridges), I cover single. We'd then still need to test the heterogeneous case, which again I couldn't test without covering all single cases anyway. So I just went with that. 

The alternative would be to expose a more granular (mockable) interface only for the purpose of testing, i.e. `prepare_single_interface(...)` which would be unit tested, and mocked when testing the whole `prepare_networking()`. Maybe that's what I should have done now that I think about it from a distance, even though I dislike tailoring interfaces to tests... Is that what you think should happen? Otherwise how do you think I should approach this concretely?

Reply:
OK, I'll see what I can do. I don't think `prepare_networking_guts` can be any simpler, but perhaps I can expose a more elementary &ndash; and mockable &ndash; interface, for testing.

Reply:
Oh yes, simpler indeed in that it offloads the body of the `for`. I think I have a path for this.

Reply:
Done, let me know what you think!

Reply:
Yeah, much better, and worth the change in implementation :)